crates/ironrdp-rdpsnd-native/src/cpal.rs

```diff
@@ -0,0 +1,244 @@
+        self.tx = None;
+    }
+}
+
+pub fn make_stream(rx_format: &AudioFormat, rx: Receiver<Vec<u8>>) -> anyhow::Result<Stream> {
+    if rx_format.format != WaveFormat::PCM {
+        bail!("Only PCM formats supported");
+    }
+    let sample_format = match rx_format.bits_per_sample {
+        8 => SampleFormat::U8,
+        16 => SampleFormat::I16,
+        _ => {
+            bail!("Only PCM 8/16 bits formats supported");
+        }
+    };
+
+    let host = cpal::default_host();
+    let device = host.default_output_device().context("No default output device")?;
+    let _supported_configs_range = device
+        .supported_output_configs()
+        .context("No supported output config")?;
+    let default_config = device.default_output_config()?;
+    debug!(?default_config);
+
+    let mut rx = RxBuffer::new(rx);
+    let config = StreamConfig {
+        channels: rx_format.n_channels,
+        sample_rate: cpal::SampleRate(rx_format.n_samples_per_sec),
+        buffer_size: cpal::BufferSize::Default,
+    };
+    debug!(?config);
+
+    let stream = device
+        .build_output_stream_raw(
+            &config,
+            sample_format,
+            move |data, _info: &cpal::OutputCallbackInfo| {
+                let data = data.bytes_mut();
+                rx.fill(data)
+            },
+            |err| error!(?err),
+            None,
+        )
+        .context("Failed to setup output stream")?;
+
+    Ok(stream)
+}
+
+struct RxBuffer {
+    receiver: Receiver<Vec<u8>>,
+    last: Option<Vec<u8>>,
+    idx: usize,
+}
+
+impl RxBuffer {
+    fn new(receiver: Receiver<Vec<u8>>) -> Self {
+        Self {
+            receiver,
+            last: None,
+            idx: 0,
+        }
+    }
+
+    fn fill(&mut self, data: &mut [u8]) {
+        let mut filled = 0;
+
+        while filled < data.len() {
+            if self.last.is_none() {
+                match self.receiver.recv_timeout(Duration::from_millis(4000)) {
+                    Ok(rx) => {
+                        debug!("{}", rx.len());
+                        self.last = Some(rx);
+                    }
+                    Err(err) => {
+                        error!(?err);
+                    }
+                }
+            }
+
+            let Some(ref last) = self.last else {
+                info!("playback rx underrun");
+                return;
+            };
+
+            #[allow(clippy::arithmetic_side_effects)]
+            while self.idx < last.len() && filled < data.len() {
+                data[filled] = last[self.idx];
+                assert!(filled < usize::MAX);
+                assert!(self.idx < usize::MAX);
+                filled += 1;
+                self.idx += 1;
+            }
+
+            // If all elements from last have been consumed, clear `self.last`
+            if self.idx >= last.len() {
+                self.last = None;
+                self.idx = 0;
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
issue: We are putting all the tests into the testsuite crate.

Reply:
The test is not meant to run under CI. It needs sound support, it's a build & interactive test for CPAL backend implementation details. I'll move it to examples/

Reply:
An example sounds appropriate :+1: